src/test/fuzz/pcp.cpp

```diff
@@ -0,0 +1,229 @@
+// Copyright (c) 2024 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <test/fuzz/FuzzedDataProvider.h>
+#include <test/fuzz/fuzz.h>
+#include <test/fuzz/util.h>
+#include <test/fuzz/util/net.h>
+
+#include <common/pcp.h>
+#include <util/check.h>
+
+using namespace std::literals;
+
+//! Fixed nonce to use in PCP port mapping requests.
+constexpr PCPMappingNonce PCP_NONCE{0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88, 0x99, 0xaa, 0xbb, 0xcc};
+
+//! NAT-PMP or PCP server port.
+constexpr uint16_t SERVER_PORT{5351};
+
+//! Number of attempts to request a NAT-PMP or PCP port mapping to the gateway.
+constexpr int NUM_TRIES{5};
+
+//! Timeout for each attempt to request a port mapping.
+constexpr std::chrono::duration TIMEOUT{100ms};
+
+/** Simple mocked socket which returns fuzzer-provided data and sometimes randomly inject some faults.
+ */
+class PCPFuzzSock final : public Sock
+{
+private:
+    // Note: we awkwardly mark all methods as const, and properties as mutable,
+    // because Sock expects all networking calls to be const.
+    mutable std::chrono::milliseconds m_time{MockableSteadyClock::INITIAL_MOCK_TIME};
+    mutable bool m_connected{false};
+    mutable CService m_bound;
+    mutable CNetAddr m_local_ip;
+    mutable CNetAddr m_gateway_ip;
+    FuzzedDataProvider& m_fuzz_data;
+
+    void ElapseTime(std::chrono::milliseconds duration) const
+    {
+        m_time += duration;
+        MockableSteadyClock::SetMockTime(m_time);
+    }
+
+public:
+    explicit PCPFuzzSock(const CNetAddr &local_ip, const CNetAddr &gateway_ip, FuzzedDataProvider& prov)
+        : Sock{INVALID_SOCKET},
+          m_local_ip(local_ip),
+          m_gateway_ip(gateway_ip),
+          m_fuzz_data(prov)
+    {
+        ElapseTime(std::chrono::seconds(0)); // start mocking steady time
+    }
+
+    PCPFuzzSock& operator=(Sock&& other) override
+    {
+        Assert(false && "Move of Sock into PCPFuzzSock not allowed.");
+        return *this;
+    }
+
+    /** Sending bytes to the gateway normally succeeds when connected. We sometimes inject some faults. */
+    ssize_t Send(const void* data, size_t len, int) const override {
+        if (!m_connected) return -1;
+        if (m_fuzz_data.ConsumeBool()) return -1;
+        return len;
+    }
+
+    /** We receive random bytes from the gateway. We make it sometimes fail, too. */
+    ssize_t Recv(void* buf, size_t len, int flags) const override
+    {
+        if (m_fuzz_data.ConsumeBool()) return -1;
+        const auto consumed_bytes{m_fuzz_data.ConsumeIntegralInRange<size_t>(0, len)};
+        const auto data{m_fuzz_data.ConsumeBytes<uint8_t>(consumed_bytes)};
+        if (!data.empty()) std::memcpy(buf, data.data(), data.size());
+        return consumed_bytes;
```

Comment:
Here's my theory for the MSan failure:

`data` might hold less bytes than `consumed_bytes` indicated because the fuzzed data provider might not have enough bytes left. This means only `data.size()` bytes of `buf` will be initialized with the memcpy but `consumed_bytes` is returned as the number of bytes "received".

This should work:

```suggestion
        const auto data{ConsumeRandomLengthByteVector(fuzzed_data_provider, len)};
        if (!data.empty()) std::memcpy(buf, data.data(), data.size());
        return data.size();
```

Or alternatively, is there a reason not use `FuzzedSock`?

Reply:
Mostly to mock `Wait` iirc. I'll reproduce your fuzz finding now and look into this. Thanks!

Reply:
Yeah, looks like i just reinvented (a poor man's) `FuzzedSock` for my target. I'll look into just using `FuzzedSock` instead tomorrow. Just confirmed your theory, returning `data.size()` instead of `consumed_bytes` fixed the crash.

Reply:
Done. It seems to have reduced the coverage somewhat. I'll run it overnight again.